src/server_key_id.rs

```diff
@@ -0,0 +1,56 @@
+//! Identifiers for homeserver signing keys used for federation.
+
+use std::{convert::TryFrom, num::NonZeroU8};
+
+use crate::{key_algorithm::KeyAlgorithm, Error};
+
+/// Key identifiers used for homeserver signing keys.
+#[derive(Clone, Debug)]
+pub struct ServerKeyId {
+    pub(crate) full_id: Box<str>,
+    pub(crate) colon_idx: NonZeroU8,
+}
+
+impl ServerKeyId {
+    /// Returns key algorithm of the server key ID.
+    pub fn algorithm(&self) -> KeyAlgorithm {
+        let algorithm_str = &self.full_id[0..self.colon_idx.get() as usize];
+        KeyAlgorithm::try_from(algorithm_str).unwrap()
+    }
+
+    /// Returns the version of the server key ID.
+    pub fn version(&self) -> &str {
+        &self.full_id[self.colon_idx.get() as usize + 1..]
+    }
+}
+
+fn try_from<S>(key_id: S) -> Result<ServerKeyId, Error>
+where
+    S: AsRef<str> + Into<Box<str>>,
+{
+    let key_str = key_id.as_ref();
+    let colon_idx = key_str
+        .find(':')
+        .ok_or(Error::MissingDelimiter)
+        .map(|index| NonZeroU8::new(index as u8).unwrap())?;
+
+    KeyAlgorithm::try_from(&key_str[0..colon_idx.get() as usize])?;
+
+    validate_version(&key_str[colon_idx.get() as usize + 1..])?;
+
+    Ok(ServerKeyId {
+        full_id: key_id.into(),
+        colon_idx,
+    })
+}
+
+common_impls!(ServerKeyId, try_from, "Key ID with algorithm and version");
+
+fn validate_version(version: &str) -> Result<(), Error> {
+    if version.is_empty() {
+        return Err(Error::MinimumLengthNotSatisfied);
+    } else if version.chars().all(|c| c.is_alphanumeric() || c == '_') {
+        return Err(Error::InvalidCharacters);
+    }
+    Ok(())
```

Comment:
I don't know why, but I feel like no newline after a closing brace line looks weird. A I being obsessive about the code style? :sweat_smile: 

```suggestion
    }

    Ok(())
```